bfvmm/include/hve/arch/intel_x64/exit_handler.h

```diff
@@ -202,71 +147,52 @@ class EXPORT_HVE exit_handler
     /// @expects none
     /// @ensures none
     ///
-    /// @param exit_handler The exit handler to handler the VM exit to
+    /// @param vcpu The vcpu this exit occured on
     ///
-    static void handle(
-        bfvmm::intel_x64::exit_handler *exit_handler) noexcept;
-
-    /// Get Host TSS
-    ///
-    /// @expects none
-    /// @ensures none
-    ///
-    /// @return Returns a pointer to the host_tss
-    ///
-    auto host_tss() noexcept
-    { return &m_host_tss; }
-
-    /// Get Host IDT
-    ///
-    /// @expects none
-    /// @ensures none
-    ///
-    /// @return Returns a pointer to the host_idt
-    ///
-    auto host_idt() noexcept
-    { return &m_host_idt; }
-
-    /// Get Host GDT
-    ///
-    /// @expects none
-    /// @ensures none
-    ///
-    /// @return Returns a pointer to the host_gdt
-    ///
-    auto host_gdt() noexcept
-    { return &m_host_gdt; }
+    /// @return True if the vmexit was successfully handled, false otherwise
+    bool handle(vcpu_t vcpu);
 
 private:
 
-    void write_host_state();
-    void write_guest_state();
-    void write_control_state();
+    // These members provide the legacy exit handler mechanism, and will be
+    // deprecated slowly
+    std::list<init_handler_delegate_t> m_exit_handlers;
+    std::list<init_handler_delegate_t> m_init_handlers;
+    std::list<fini_handler_delegate_t> m_fini_handlers;
+    std::array<std::list<handler_delegate_t>, 128> m_exit_handlers_array;
+
+    // These members will provide the new exit handler mechanism
+    std::array<handler_delegate_t, 128> m_exit_delegates;
```

Comment:
why do we have two arrays?

Reply:
One array is for the "legacy" mechanism (i.e. the current EAPIS style handlers) and the second array is for the new delegator based mechanism. They both exists so that we can slowly deprecate one exit reason at a time.